### PR TITLE
Marking Cargo.lock as generated

### LINKED ISSUE #6180: Marking Cargo.lock as generated
Our code review tool (phabricator) will automatically hide changes to files that are marked as generated (via detecting the text `@generated` in the file)

It would be convenient if Cargo.lock had `// @generated` in the file, or some customizable comment to this effect.

I *think* we can configure phabricator to avoid this issue, but it seems also reasonable to have Cargo's defaults play nicely with phabricator's defaults.

### PR DESCRIPTION
**Pull request by dwijnand**
12 weeks have passed, let's mark lock files as generated!

Fixes #6180

### CODE REVIEW

**Comment by alexcrichton** [comment]
Thanks! 

I wonder if we could perhaps expand the comment a bit? Something along the lines of adding human-readable text about how it's a generated file and should not be edited, followed by `@generated` perhaps? I think it'd also be good to have a comment in the Cargo source indicating what `@generated` means and why we're emitting it.

**Comment by dwijnand** [comment]
Sure!  I was a little unsure, so I pushed my attempt.  WDYT?  I'm happy to keep tweaking it.

**Comment by dwijnand** [comment]
I used https://github.com/github/linguist/blob/f1f4018d7a9950c1d9cf3d0887d994cf496d63c4/lib/linguist/generated.rb as inspiration (note Cargo.lock is already ignored there).  Specifically, I preferred "Not intended for manual editing" over "Do not edit" (or "DO NOT EDIT").

File: src/cargo/ops/lockfile.rs [hunk 1]
```diff
@@ -43,13 +43,21 @@ pub fn write_pkg_lockfile(ws: &Workspace<'_>, resolve: &Resolve) -> CargoResult<
...
+    // Specifically Phabricator ignores files containing "@generated", so we use that.
+    out.push_str("# This file is automatically @generated by Cargo.  Not intended for manual editing.\n");
+    // and preserve any other top comments
     if let Ok(orig) = &orig {
-        for line in orig.lines().take_while(|line| line.starts_with('#')) {
-            out.push_str(line);
-            out.push_str("\n");
+        let mut comments = orig.lines().take_while(|line| line.starts_with('#'));
+        if let Some(first) = comments.next() {
+            if first != "# This file is automatically @generated by Cargo.  Not intended for manual editing." {
```
Comment (alexcrichton) [inline src/cargo/ops/lockfile.rs]: Could this be deduplicated with the comment above? Additionally, could a line break be in here to ensure it's under 80 columns?

**Comment by alexcrichton** [comment]
This is going to affect every single Rust project and likely surprise a lot of people. I think we're ready for this change though and the surprise is mostly going to be a result of change, not that it's a change that shouldn't have happened. Regardless I'd like to make sure we're all on board with this, so..


@rfcbot fcp merge

File: src/cargo/ops/lockfile.rs (hunk 1 above)
Comment (dwijnand) [reply src/cargo/ops/lockfile.rs]: > Could this be deduplicated with the comment above?

Doh, good call.

> Additionally, could a line break be in here to ensure it's under 80 columns?

I cheated and made it 1 line, under 80 columns.

**Comment by ehuss** [comment]
> This is going to affect every single Rust project

Just to be clear, the only effect is that if you check in the Cargo.lock into VCS, it will show up as dirty after the first build with a new cargo?  If you don't check the lock file in, you probably won't even notice?

**Comment by alexcrichton** [comment]
@ehuss an excellent clarification, and spot on

File: src/cargo/ops/lockfile.rs [hunk 2]
```diff
@@ -43,13 +43,23 @@ pub fn write_pkg_lockfile(ws: &Workspace<'_>, resolve: &Resolve) -> CargoResult<
 
     let mut out = String::new();
 
-    // Preserve the top comments in the lockfile
-    // This is in preparation for marking it as generated
-    // https://github.com/rust-lang/cargo/issues/6180
+    // At the start of the file we notify the reader that the file is generated.
+    // Specifically Phabricator ignores files containing "@generated", so we use that.
+    let marker_line = "# This file is automatically @generated.  Not intended for manual editing.";
```
Comment (ln-nl) [inline src/cargo/ops/lockfile.rs]: Excuse me, why there is two spaces here?

Comment (dwijnand) [reply src/cargo/ops/lockfile.rs]: https://en.wikipedia.org/wiki/Sentence_spacing

**Comment by lukaslueg** [comment]
Maybe nitpicking but could we have a full English sentence instead of "Not intended for manual editing" here? E.g. "This file is not intended for manual editing" or "It is not ...". Also, if this is an auto-generated egg, where is the chicken - who generated it?

`This file was automatically @generated by Cargo.  It is not intended for manual editing.`

**Comment by alexcrichton** [comment]
@dwijnand want to follow up with @lukaslueg's thoughts and we can merge?

**Comment by dwijnand** [comment]
I thought it was fine, but I don't mind tweaking it.  I dropped the "automatically" because I want to keep it one line, under 80 chars, as looking in https://github.com/github/linguist/blob/f1f4018d7a9950c1d9cf3d0887d994cf496d63c4/lib/linguist/generated.rb it's probably best, over multi-line markers.

**Comment by dwijnand** [comment]
Yes, it would appear so, from that linguist file.

But I realised we can just call the first one the marker line.  So it's now two lines:

```toml
# This file is automatically @generated by Cargo.
# It is not intended for manual editing.
```

**Pull request merged**